### DISCUSSION TITLE
State memory handling

### DISCUSSION THREAD

**Discussion by mattldawson** [NCAR/micm#648] [Ideas]
After moving temporary solver variables to the [`State`](https://github.com/NCAR/micm/blob/41133753bedecb41a13f5d719744872aefeb166b/include/micm/solver/state.hpp#L36-L94) class, we should consider options for optionally allocating/deallocating these temporary variables.

This could include:
- the ability to create "minimal" instances of the Matrix classes (without index arrays for the sparse matrices and without host memory for CUDA matrices, e.g.)
- maybe a function on the `State` class to allocate/deallocate all temporary variables to reduce the memory footprint between calls to `Solve`.

**Comment by sjsprecious** [comment]
According to the discussion with @mattldawson , one way to do this is to create another simple constructor for the `CudaDenseMatrix` and `VectorMatrix`, which only allocates the memory space of a given dimension matrix without allocating additional index arrays. The constructor may look like `MatrixType(Bool minimal_instance, int x_dim, int y_dim)`?

**Comment by sjsprecious** [comment]
Also we can implement a separate `CleanUp` function in the `State` class to manually deallocate the temporary variables if necessary.
